File: Algorithms/1. Sorting/quick-sort/quickSortLomuto.py
""" Lomuto partition scheme Quick Sort"""

# partition the array
def partition(array, left, right):
  pivot = array[right] # choose rightmost element as pivot
  index = left # swap index be the start element

  # iterate the element and swap elements < pivot
  for j in range(left, right):
    if array[j] < pivot:
      array[index], array[j] = array[j], array[index]
      index += 1 # increment next index for swapping

  array[index], array[right] = array[right], array[index]
  return index

# divide array into halves
def quick_sort(array, left, right):
  if left < right:
    pivot = partition(array, left, right)
    quick_sort(array, left, pivot - 1)
    quick_sort(array, pivot + 1, right)

# testing the algorithm
print("Lomuto partition scheme------")
print("Quick sort O(nlogn) algorithm")

# array before sorting
array = [9, 8, 7, 6, 5, 4, 3, 2, 1]
print("Array before sorting")
print(array)

# test the quick sort algorithm
print("Array after sorting")
quick_sort(array, 0, len(array)-1)
print(array)

File: Algorithms/1. Sorting and Selection/quick-sort/quickSortLomuto.cpp
// Lomuto partition scheme Quick Sort
#include <iostream>
#include <vector>
using namespace std;

// swap function
void swap(int* num1, int* num2) {
  int temp = *num1;
  *num1 = *num2;
  *num2 = temp;
}

// Lomuto partition scheme
int partition(vector<int>& array, int left, int right) {
  int pivot = array[right]; // choose rightmost element for comparison
  int index = left; // swap element index

  // swap elements < pivot with current element in a loop
  for (int j = left; j < right; j++) {
    if (array[j] < pivot) {
      swap(array[index], array[j]);
      index++; // increment next index for swapping
    }
  }
  swap(array[index], array[right]);
  return index;
}

// divide the array and sort it
void quickSort(vector<int>& array, int left, int right) {
  if (left < right) {
    int pivot = partition(array, left, right);
    quickSort(array, left, pivot-1);
    quickSort(array, pivot+1, right);
  }
}

int main() {
  cout << "Lomuto partition scheme------\n";
  cout << "Quick sort O(nlogn) algorithm\n";
  vector<int> array{5,4,3,2,1};
  cout << "Before sorting: ";
  for (auto num: array)
    cout << num << " ";
  cout << endl;

  // run quicksort algorithm
  quickSort(array, 0, array.size()-1);
  cout << "After sorting: ";
  for (auto num: array)
    cout << num << " ";
  cout << endl;
}